metisp/pymetis/src/pymetis/recipes/metis_det_dark.py

```diff
@@ -77,20 +78,276 @@ class GainMapInput(OptionalInputMixin, GainMapInput):
+        header.append(cpl.core.Property("QC DARK RMS", cpl.core.Type.DOUBLE,
+                                        qcrms, "[ADU] rms value of master dark"))
+        header.append(cpl.core.Property("QC DARK NBADPIX", cpl.core.Type.DOUBLE,
+                                        qcnbad, "[ADU] number of bad pixels"))
+        header.append(cpl.core.Property("QC DARK NCOLDPIX", cpl.core.Type.DOUBLE,
+                                        qcncold, "[ADU] number of cold pixels"))
+        header.append(cpl.core.Property("QC DARK NHOTPIX", cpl.core.Type.DOUBLE,
+                                        qcnhot, "[ADU] number of hot pixels"))
+        header.append(cpl.core.Property("QC DARK MEDIAN MEAN", cpl.core.Type.DOUBLE,
+                                        qcmedmean, "[ADU] median value of mean values of individual input images"))
+        header.append(cpl.core.Property("QC DARK MEDIAN MEDIAN", cpl.core.Type.DOUBLE,
+                                        qcmedmed, "[ADU] median value of median values of individual input images"))
+        header.append(cpl.core.Property("QC DARK MEDIAN RMS", cpl.core.Type.DOUBLE,
+                                        qcmedrms, "[ADU] median value of RMS values of individual input images"))
+        header.append(cpl.core.Property("QC DARK MEDIAN MIN", cpl.core.Type.DOUBLE,
+                                        qcmedmin, "[ADU] median value of min values of individual input images"))
+        header.append(cpl.core.Property("QC DARK MEDIAN MAX", cpl.core.Type.DOUBLE,
+                                         qcmedmax, "[ADU] median value of max values of individual input images"))
+            
+        product = self.ProductMasterDark(header, combined_image,)
+        
         return {product}
 
 
+    def metis_bpm_3d_compute(self,imageList,kappaLow, kappaHigh):
+
+        """Calculate outlier pixels based on high/low thresholds based on the frame to frame variation of a pixel"""
+
+        imsum = cpl.core.Image.zeros_like(imageList[0])
+        im2sum = cpl.core.Image.zeros_like(imageList[0])
+        
+        for im in imageList:
+            imTem = cpl.core.Image.zeros_like(im)
+            imTem.copy_into(im, 0, 0)
+            imsum.add(im)
+            im.power(2)
+            im2sum.add(im)
+
+        imsum.divide_scalar(len(imageList))
+        imsum.power(2)
+        im2sum.divide_scalar(len(imageList))
+
+        imsum.add(im2sum)
+        imsum.power(0.5)
+
+        imMed = imsum.get_median()
+        imRMS = imsum.get_stdev()
+        
+        mask = cpl.core.Mask.threshold_image(imsum, imMed - kappaLow*imRMS, imMed + kappaHigh*imRMS, 1)
+        qcnbad = mask.count()
+        mask = cpl.core.Image(mask,dtype = cpl.core.Type.INT)
+
+        return mask, qcnbad
+        
+
+
+        
+                             
+        
+        
+        
+    def collapse_raw_with_error(self,imageList, method, readNoise):
```

Comment:
We should at some point refactor these functions so they can easily be reused. Starting with putting the logic in a function like this is a good first step.